lib/natalie/compiler/pass0c.rb

```diff
@@ -126,43 +160,118 @@ def process_call(exp, block_with_args = nil)
               end
             end
           elsif args.length == 0
-            if method == :!
-              # This can fail and fallback, so we should not try to process it again
-              return canary_const_not(exp, values)
+            # These can fail and fallback, so we should not try to process them again
+            case method
+            when :!
+              return canary_const_not(exp)
+            when :is_truthy
+              return canary_const_truthy(exp)
             end
           end
         end
         return exp
       end
 
-      def canary_array_min(exp, values)
+      def canary_array_min(exp, values, args = nil, function = nil)
         return exp.new(:nil) if values.empty?
         return values[0] if values.length == 1
         first, second, *rest = values
+        if args.nil? && function.nil?
+          return(
+            exp.new(
+              :if,
+              exp.new(:call, first, :<, second),
+              canary_array_min(exp, [first, *rest]),
+              canary_array_min(exp, [second, *rest]),
+            )
+          )
+        end
+        return exp if args.length != 3 # Wat
+
+        new_function1 = function.map { |expr| ((expr == s(:lvar, args[1])) ? first : expr) }
+        new_function2 = function.map { |expr| ((expr == s(:lvar, args[2])) ? second : expr) }
```

Comment:
I have no idea what this is doing. Needs comments or a named method to describe it.

Reply:
I will add a comment about this, maybe even outline it and rename the variables

But for now,
we replace every mention of the variable passed into the block (ie `s(:lvar, args[1])`) with the actual value we are currently iterating on